### PR TITLE
use rclcpp::Node for generic pub/sub

### PR DESCRIPTION
**Pull request by Karsten1987**
There were some functions for topic expansion and filtering on the `rosbag2node` which I moved into a new class `TopicFilter`.

### CODE REVIEW

File: rosbag2_transport/src/rosbag2_transport/topic_filter.hpp [hunk 1]
```diff
@@ -12,31 +12,33 @@
...
-
-#include "rclcpp/rclcpp.hpp"
+#include <unordered_map>
+#include <vector>
 
 namespace rosbag2_transport
 {
 
-class GenericPublisher : public rclcpp::PublisherBase
+class TopicFilter
```
Comment (emersonknapp) [inline rosbag2_transport/src/rosbag2_transport/topic_filter.hpp:26]: nitpick: could this just be a namespace? do you have thoughts about it being stateful in the future?

File: rosbag2_transport/src/rosbag2_transport/topic_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      ROSBAG2_TRANSPORT_LOG_ERROR_STREAM(
+        "Topic '" << topic_and_type.first <<
+          "' has several types associated. Only topics with one type are supported");
+      continue;
+    }
+
+    // According to rclpy's implementation, the indicator for a hidden topic is a leading '_'
+    // https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/topic_or_service_is_hidden.py#L15
+    if (!include_hidden_topics) {
+      auto tokens = rcpputils::split(topic_and_type.first, '/', true);  // skip empty
```
Comment (emersonknapp) [inline rosbag2_transport/src/rosbag2_transport/topic_filter.cpp:69]: Slightly more clear as

```suggestion
      auto tokens = rcpputils::split(topic_and_type.first, '/', true /* skip_empty */);
```

File: rosbag2_transport/src/rosbag2_transport/topic_filter.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
```
Comment (emersonknapp) [inline rosbag2_transport/src/rosbag2_transport/topic_filter.cpp:1]:
```suggestion
// Copyright 2021, Bosch Software Innovations GmbH.
```

?

File: rosbag2_transport/src/rosbag2_transport/topic_filter.hpp [hunk 4]
```diff
@@ -12,31 +12,33 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#ifndef ROSBAG2_TRANSPORT__GENERIC_PUBLISHER_HPP_
-#define ROSBAG2_TRANSPORT__GENERIC_PUBLISHER_HPP_
+#ifndef ROSBAG2_TRANSPORT__TOPIC_FILTER_HPP_
```
Comment (emersonknapp) [inline rosbag2_transport/src/rosbag2_transport/topic_filter.hpp:15]: ^^^ copyright 2021?

File: rosbag2_transport/test/rosbag2_transport/test_topic_filter.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
```
Comment (emersonknapp) [inline rosbag2_transport/test/rosbag2_transport/test_topic_filter.cpp:1]:
```suggestion
// Copyright 2021, Bosch Software Innovations GmbH.
```

?

File: rosbag2_transport/test/rosbag2_transport/test_topic_filter.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+using namespace ::testing;  // NOLINT
+
+TEST(TestTopicFilter, filter_topics_with_more_than_one_type) {
+  std::map<std::string, std::vector<std::string>> topic_with_type;
+  topic_with_type.insert({"topic/a", {"type_a"}});
+  topic_with_type.insert({"topic/b", {"type_b"}});
+  topic_with_type.insert({"topic/c", {"type_c"}});
+  {
+    auto filtered_topics = rosbag2_transport::TopicFilter::filter_topics_with_more_than_one_type(
+      topic_with_type, true);  // include hidden topics
```
Comment (emersonknapp) [inline rosbag2_transport/test/rosbag2_transport/test_topic_filter.cpp:35]: As I put in another comment - I prefer the style:

```suggestion
      topic_with_type, true /* include_hidden_topics */); 
```

Up to you if you want to follow that convention

**Review by emersonknapp — APPROVED** [review]
No major complaints - this is a satisfying diff :)

**Comment by Karsten1987** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14257/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9049/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11941/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14357/)

**Comment by Karsten1987** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14261/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9052/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11944/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14360/)

**Comment by Karsten1987** [comment]
Windows warning is unrelated, comes from cyclonedds.

**Pull request merged**